src/ocean/util/container/btree/BTreeMap.d

```diff
@@ -0,0 +1,664 @@
+      element(i), and all the elements in a subtree pointed by p(i) are larger
+      than a element(i).
+
+    - Every node may contain at most 2*degree - 1 elements. Therefore, an internal
+      node may be root for at most 2*degree subtrees.
+
+    - Every node other than root must have at least (degree-1) elements.
+
+    - Every internal node other than the root has at least degree children.
+
+    - Every node has the following attributes:
+
+        node.number_of_elements = number of elements node currently contains
+            (as nodes are statically allocated block, the number of the elements
+             may not necessarily be maximal.
+        node.elements = the elements themselves
+        node.is_leaf = indicator if the node is a leaf.
+        node.child_nodes = number_of_elements+1 pointers to the subtrees whose
+            roots are this node. We refer to these nodes as child-nodes (this
+            node is a direct parent of them).
+
+    - All leaves have the same depth
+
+    Other than insert and delete operations, inorder traversal method is provided,
+    which traverses the tree in inorder order (left subtree is visited first,
+    then the element separating these subtrees, then the right subtree, then
+    the next element, etc.). This provides sequential access to the tree.
+    There's a recursive and range-based implementation of the inorder
+    traversal.
+
+    Range-based traversal is supported through `BTreeMapRange` structure, found in
+    (and here publicly imported) in ocean.util.container.btree.Btree.
+
+    CPU complexity of the operations (n is number of the elements, t is degree,
+    h is the tree height):
+
+        - Searching: O(th) = O(t * log_t(n))
+        - Inserting: O(th) = O(t * log_t(n))
+        - Deleting:  O(th) = O(t * log_t(n))
+
+    Complexity of the memory access (to read/write the node from/in memory):
+
+        - Searching O(h) = O(log_t(n))
+        - Inserting O(h) = O(log_t(n))
+        - Deleting: O(h) = O(log_t(n))
+
+    References:
+        - https://en.wikipedia.org/wiki/B-tree
+        - Thomas H. Cormen et al. "Introduction to Algorithms, 3rd edition"
+
+    Params:
+        KeyType = type of the key.
+        TreeElementType = type of the element to store.
+        tree_degree = degree of the tree (refer to the documentation)
+
+******************************************************************************/
+
+struct BTreeMap(TreeKeyType, TreeElementType, int tree_degree)
+{
+    import ocean.util.container.btree.Implementation;
+
+    static assert (tree_degree > 0);
+
+    /**************************************************************************
+
+        Constructor.
+
+        Params:
+            allocator = memory manager to use to allocate/deallocate memory
+
+    ***************************************************************************/
+
+    package void init (IMemManager allocator = mallocMemManager)
```

Comment:
It is not a good idea to name method `init` because of paren-less calls and built-in `T.init` defined for all types.